File: src/Web/Dart/factory_method.dart
abstract interface class Database {
  void connect();
  void query();
}

final class PostgresDatabase implements Database {
  @override
  void connect() => print('PostgreSQL connect');

  @override
  void query() => print('PostgreSQL query');
}

final class MySqlDatabase implements Database {
  @override
  void connect() => print('MySQL connect');

  @override
  void query() => print('MySQL query');
}

abstract class DatabaseCreator {
  Database createDatabase();

  void useDatabase() {
    final database = createDatabase();
    database.connect();
    database.query();
  }
}

final class PostgresCreator extends DatabaseCreator {
  @override
  Database createDatabase() => PostgresDatabase();
}

final class MySqlCreator extends DatabaseCreator {
  @override
  Database createDatabase() => MySqlDatabase();
}

void main() {
  PostgresCreator().useDatabase();
  MySqlCreator().useDatabase();
}

File: src/Web/TypeScriptTS/factory_method.ts
interface Database {
  connect(): void;
  query(): void;
}

class PostgresDatabase implements Database {
  connect(): void { console.log("PostgreSQL connect"); }
  query(): void { console.log("PostgreSQL query"); }
}

class MySqlDatabase implements Database {
  connect(): void { console.log("MySQL connect"); }
  query(): void { console.log("MySQL query"); }
}

abstract class DatabaseCreator {
  protected abstract createDatabase(): Database;

  useDatabase(): void {
    const database = this.createDatabase();
    database.connect();
    database.query();
  }
}

class PostgresCreator extends DatabaseCreator {
  protected createDatabase(): Database { return new PostgresDatabase(); }
}

class MySqlCreator extends DatabaseCreator {
  protected createDatabase(): Database { return new MySqlDatabase(); }
}

new PostgresCreator().useDatabase();
new MySqlCreator().useDatabase();
